tests/rptest/tests/pandaproxy_test.py

```diff
@@ -2129,3 +2132,130 @@ def test_upgrade_and_enable_basic_auth(self, base_release: tuple[int, int],
+                size=self.bytes_per_second)
+            data = {"records": [{"value": msg_ascii}]}
+            self.logger.debug(f"Producing {json.dumps(data)}")
+            self.result_raw = self._produce_topic(
+                self.topic.name,
+                json.dumps(data),
+                headers=HTTP_PRODUCE_JSON_V2_TOPIC_HEADERS,
+                auth=self.auth_tuple)
+            assert self.result_raw.status_code == requests.codes.ok
+            produce_result = self.result_raw.json()
+            print(f"produce result {json.dumps(produce_result)}")
+            self.logger.debug(f"produce result {json.dumps(produce_result)}")
+            assert "offsets" in produce_result
+            assert len(produce_result["offsets"]) == 1
+            offset = produce_result["offsets"][0]
+            assert "error_code" not in offset, offset["error_code"]
+            if offset["partition"] not in self.partition_offsets:
+                self.partition_offsets[offset["partition"]] = {}
+            self.partition_offsets[offset["partition"]][
+                offset["offset"]] = msg_ascii
+
+            return False
+
+        try:
+            wait_until(do_produce, timeout_sec=self.timeout_sec, backoff_sec=1)
+        except ducktape.errors.TimeoutError as ex:
+            self.done = True
+        except Exception as ex:
+            self.exception = ex
+
+
+class VerifiablePandaproxyConsumer(threading.Thread):
+    def __init__(self,
+                 topic: TopicSpec,
+                 consumer: Consumer,
+                 verifiable_producers: list[VerifiablePandaproxyProducer],
+                 timeout_sec: int,
+                 logger: TestLoggerMaker,
+                 auth_tuple: Optional[tuple[str, str]] = None):
+        threading.Thread.__init__(self)
+        self.daemon = True
+        self.topic = topic
+        self.consumer = consumer
+        self.verifiable_producers = verifiable_producers
+        self.timeout_sec = timeout_sec
+        self.logger = logger
+        self.auth_tuple = auth_tuple
+        self.result_raw = None
+        self.done = False
+        self.exception = None
+
+    def _filter_producer_by_partition_offset(self, partition: int,
+                                             offset: int):
+        try:
+            prod = next(
+                filter(
+                    lambda prod: partition in prod.partition_offsets and offset
+                    in prod.partition_offsets[partition],
+                    self.verifiable_producers))
+            return prod
+        except StopIteration:
+            raise RuntimeError(
+                f"Failed to find producer with partition-offset {partition}-{offset}"
+            )
+
+    def _fetch_message_by_partition_offset(self, partition: int, offset: int):
+        verifiable_producer = self._filter_producer_by_partition_offset(
+            partition, offset)
+        return verifiable_producer.partition_offsets[partition][offset]
```

Comment:
```suggestion
    def _fetch_message_by_partition_offset(self, partition: int, offset: int):
        try:
            return self.verifiable_producer.partition_offsets[partition][
                offset]
        except KeyError:
            raise RuntimeError(
                f"Failed to find producer with partition-offset {partition}-{offset}"
            )
```

Reply:
I may be misunderstanding something here, but the verifiable producers are stored in a list and each consumer needs find the producer that wrote to the appropriate partition-offset.

Neverthless, this may change when I refactor the producer and consumer

Reply:
This changed, see the VerifiablePriorityQueue